opendevin/sandbox/docker.py

```diff
@@ -0,0 +1,145 @@
+import os
+import pty
+import sys
+import uuid
+import time
+import shlex
+import select
+import subprocess
+from typing import List
+from collections import namedtuple
+
+InputType = namedtuple("InputDtype", ["content"])
+OutputType = namedtuple("OutputDtype", ["content"])
+
+
+class DockerInteractive:
+    CONTAINER_IMAGE = "opendevin/sandbox:latest"
+
+    def __init__(
+        self,
+        workspace_dir: str = None,
+        container_image: str = None,
+        timeout: int = 5
+    ):
+        self.instance_id: str = uuid.uuid4()
+        if workspace_dir is not None:
+            assert os.path.exists(workspace_dir), f"Directory {workspace_dir} does not exist."
+            # expand to absolute path
+            workspace_dir = os.path.abspath(workspace_dir)
+        else:
+            workspace_dir = os.getcwd()
+            print(f"workspace unspecified, using current directory: {workspace_dir}")
+        
+        # TODO: this timeout is actually essential - need a better way to set it
+        # if it is too short, the container may still waiting for previous
+        # command to finish (e.g. apt-get update)
+        # if it is too long, the user may have to wait for a unnecessary long time
+        self.timeout: int = timeout
+
+        if container_image is None:
+            container_image = self.CONTAINER_IMAGE
```

Comment:
This is super helpful--I imagine folks are going to want to define custom images so that the LLM doesn't have to e.g. install nodejs or rust every time it starts a new task.

We'll probably want to make this configurable in the UI.